File: apps/elixir/lib/algorithm/trees.ex
defmodule Algorithm.BinarySearchTree do
  @moduledoc "二分探索木（イミュータブル実装）"

  defstruct [:value, :left, :right]

  @doc "空の BST を作成"
  def new, do: nil

  @doc "値を挿入"
  def insert(nil, value), do: %__MODULE__{value: value, left: nil, right: nil}

  def insert(%__MODULE__{value: v} = tree, value) when value < v,
    do: %{tree | left: insert(tree.left, value)}

  def insert(%__MODULE__{value: v} = tree, value) when value > v,
    do: %{tree | right: insert(tree.right, value)}

  def insert(tree, _value), do: tree

  @doc "値の存在確認"
  def member?(nil, _value), do: false
  def member?(%__MODULE__{value: v}, value) when v == value, do: true

  def member?(%__MODULE__{value: v, left: left}, value) when value < v,
    do: member?(left, value)

  def member?(%__MODULE__{right: right}, value), do: member?(right, value)

  @doc "ノードの削除"
  def delete(nil, _value), do: nil

  def delete(%__MODULE__{value: v, left: l, right: r}, value) when value < v,
    do: %__MODULE__{value: v, left: delete(l, value), right: r}

  def delete(%__MODULE__{value: v, left: l, right: r}, value) when value > v,
    do: %__MODULE__{value: v, left: l, right: delete(r, value)}

  def delete(%__MODULE__{value: _v, left: nil, right: r}, _value), do: r
  def delete(%__MODULE__{value: _v, left: l, right: nil}, _value), do: l

  def delete(%__MODULE__{value: _v, left: l, right: r}, _value) do
    min_val = find_min(r)
    %__MODULE__{value: min_val, left: l, right: delete(r, min_val)}
  end

  @doc "最小値を取得"
  def find_min(nil), do: nil
  def find_min(%__MODULE__{left: nil, value: v}), do: v
  def find_min(%__MODULE__{left: l}), do: find_min(l)

  @doc "最大値を取得"
  def find_max(nil), do: nil
  def find_max(%__MODULE__{right: nil, value: v}), do: v
  def find_max(%__MODULE__{right: r}), do: find_max(r)

  @doc "中順走査（昇順）"
  def in_order(nil), do: []

  def in_order(%__MODULE__{value: v, left: l, right: r}),
    do: in_order(l) ++ [v] ++ in_order(r)

  @doc "前順走査"
  def pre_order(nil), do: []

  def pre_order(%__MODULE__{value: v, left: l, right: r}),
    do: [v] ++ pre_order(l) ++ pre_order(r)

  @doc "後順走査"
  def post_order(nil), do: []

  def post_order(%__MODULE__{value: v, left: l, right: r}),
    do: post_order(l) ++ post_order(r) ++ [v]
end

File: apps/clojure/src/algorithm/trees.clj
(ns algorithm.trees
  "第9章 木構造（二分探索木）")

(defrecord BSTNode [key left right])

(defn bst-insert
  "二分探索木にキーを挿入する（不変データ構造版）"
  [tree key]
  (cond
    (nil? tree) (->BSTNode key nil nil)
    (= key (:key tree)) tree ;; 重複は無視
    (< key (:key tree)) (->BSTNode (:key tree) (bst-insert (:left tree) key) (:right tree))
    :else (->BSTNode (:key tree) (:left tree) (bst-insert (:right tree) key))))

(defn bst-search
  "二分探索木からキーを検索する"
  [tree key]
  (cond
    (nil? tree) false
    (= key (:key tree)) true
    (< key (:key tree)) (recur (:left tree) key)
    :else (recur (:right tree) key)))

(defn- find-min
  "部分木の最小ノードを見つける"
  [tree]
  (if (nil? (:left tree))
    tree
    (recur (:left tree))))

(defn bst-delete
  "二分探索木からキーを削除する（不変データ構造版）"
  [tree key]
  (cond
    (nil? tree) nil
    (< key (:key tree)) (->BSTNode (:key tree) (bst-delete (:left tree) key) (:right tree))
    (> key (:key tree)) (->BSTNode (:key tree) (:left tree) (bst-delete (:right tree) key))
    ;; key == (:key tree): 削除対象
    :else
    (cond
      ;; 葉ノード
      (and (nil? (:left tree)) (nil? (:right tree))) nil
      ;; 左子のみ
      (nil? (:right tree)) (:left tree)
      ;; 右子のみ
      (nil? (:left tree)) (:right tree)
      ;; 子が2つ: 右部分木の最小値で置き換え
      :else
      (let [successor (find-min (:right tree))]
        (->BSTNode (:key successor) (:left tree) (bst-delete (:right tree) (:key successor)))))))

(defn in-order
  "中順走査（昇順）"
  [tree]
  (if (nil? tree)
    []
    (concat (in-order (:left tree))
            [(:key tree)]
            (in-order (:right tree)))))

(defn pre-order
  "前順走査"
  [tree]
  (if (nil? tree)
    []
    (concat [(:key tree)]
            (pre-order (:left tree))
            (pre-order (:right tree)))))

(defn post-order
  "後順走査"
  [tree]
  (if (nil? tree)
    []
    (concat (post-order (:left tree))
            (post-order (:right tree))
            [(:key tree)])))
